### PR TITLE
Specter2: Style validation

### PR DESCRIPTION
**Pull request by lynnerang**
Consumes the new 'library' metadata property applied to DS styles via Stapler and uses it to show a validation error if any conflict with the frame library.

### CODE REVIEW

File: specter2/src/App.ts [hunk 1]
```diff
@@ -578,7 +575,7 @@ export default class App {
           return;
         }
       } else {
-        const isTopFrameSelected: boolean = !!selection.find(({ type, parent }) => [
+        const isTopFrameSelected: boolean = selection.some(({ type, parent }) => [
```
Comment (gblakeman) [inline specter2/src/App.ts:578]: Lol, I don’t think I even knew about `some` 😂

File: specter2/src/Identifier.ts [hunk 2]
```diff
@@ -249,37 +249,51 @@ export default class Identifier {
...
         textStyleId: this.node.textStyleId,
         textAlignHorizontal: this.node.textAlignHorizontal,
       });
 
-      if (textToSet) {
-        styleText = `${textToSet}${subtextToSet ? `, ${subtextToSet}` : ''}`;
+      if (text) {
+        // Note: stroke and fill styles both point to the COLOR/PaintStyle details
+        library = styleIds.reduce((acc, key) => {
+          if (![undefined, figma.mixed].includes(this.node[key])) {
```
Comment (gblakeman) [inline specter2/src/Identifier.ts:283]: This is used a couple times and is pretty dense (hard to follow) – maybe worth setting it as its own variable separately?

File: specter2/src/Identifier.ts [hunk 3]
```diff
@@ -249,37 +249,51 @@ export default class Identifier {
...
         strokeStyleId: this.node.strokeStyleId,
         textStyleId: this.node.textStyleId,
         textAlignHorizontal: this.node.textAlignHorizontal,
       });
 
-      if (textToSet) {
-        styleText = `${textToSet}${subtextToSet ? `, ${subtextToSet}` : ''}`;
+      if (text) {
+        // Note: stroke and fill styles both point to the COLOR/PaintStyle details
+        library = styleIds.reduce((acc, key) => {
```
Comment (gblakeman) [inline specter2/src/Identifier.ts:282]: What does `acc` stand for?

**Review by gblakeman — APPROVED** [review]
🌈🦄

File: specter2/src/Identifier.ts (hunk 2 above)
Comment (lynnerang) [reply specter2/src/Identifier.ts:283]: yeahh I'm with you.  we do it in a few places, might just extract it out into a shorter named function

File: specter2/src/Identifier.ts (hunk 3 above)
Comment (lynnerang) [reply specter2/src/Identifier.ts:282]: "accumulator", it's a standard reference for the reduce method first param

File: specter2/src/App.ts [hunk 4]
```diff
@@ -120,10 +120,7 @@ const getViewData = (
 
     if (node) {
       const nodeData: NodeData = new NodeData(getData(node, 'nodeData'));
-      const isSelected: boolean = !![
-        node.id,
-        annotationId,
-      ].find((nId) => selectedIds.includes(nId));
+      const isSelected: boolean = [node.id, annotationId].some((nId) => selectedIds.includes(nId));
```
Comment (joeyyang) [inline specter2/src/App.ts:123]: I think this pattern is really clever if we're checking a lot of values, but ultimately we're just checking two conditions here, right? If so, consider making those checks more readable, like:
```ts
const isSelected: boolean = (selectedIds.includes(node.id) || selectedIds.includes(annotationId));
```

File: specter2/src/views/Annotations/Annotations.svelte [hunk 5]
```diff
@@ -54,7 +54,8 @@
       isMissingA11yName
       || (role === 'image' && !labels.alt)
       || hasPlatformConflict(item.platform, platform)
-      || (item.library && libraries && !libraries.includes(item.library))
+      || (item.component && item.library && libraries && !libraries.includes(item.library))
+      || (item.style && item.styleLibrary && libraries && !libraries.includes(item.styleLibrary))
```
Comment (joeyyang) [inline specter2/src/views/Annotations/Annotations.svelte:58]: this check is fairly complex - can we give each condition a name for readability? i.e., 
```ts
const imageHasNoAltTag = (role === 'image' && !labels.alt);
const itemHasPlatformConflict = (item.platform, platform);
const itemIsNotInLibrary = (item.component && item.library && libraries && !libraries.includes(item.library));
const styleIsNotInLibrary = (item.style && item.styleLibrary && libraries && !libraries.includes(item.styleLibrary));
...

if (
  isMissingA11yName
  || imageHasNoAltTag
  || itemHasPlatformConflict
  || itemIsNotInLibrary
  || styleIsNotInLibrary
)
```

File: specter2/src/Identifier.ts [hunk 6]
```diff
@@ -249,37 +249,51 @@ export default class Identifier {
...
         strokeStyleId: this.node.strokeStyleId,
         textStyleId: this.node.textStyleId,
         textAlignHorizontal: this.node.textAlignHorizontal,
       });
 
-      if (textToSet) {
-        styleText = `${textToSet}${subtextToSet ? `, ${subtextToSet}` : ''}`;
+      if (text) {
+        // Note: stroke and fill styles both point to the COLOR/PaintStyle details
+        library = styleIds.reduce((acc, id) => {
```
Comment (joeyyang) [inline specter2/src/Identifier.ts:282]: could we use a more descriptive name for `acc`, perhaps `libraryName` or `foundLibraryName`?

**Review by joeyyang — APPROVED** [review]
Mostly readability stuff but this looks good!

File: specter2/src/Identifier.ts [hunk 7]
```diff
@@ -249,37 +249,51 @@ export default class Identifier {
...
+    let library: Library;
 
-    const hasStyles: boolean = [
-      'effect',
-      'fill',
-      'stroke',
-      'text',
-    ].find((key) => this.node[`${key}StyleId`]) !== undefined;
+    const frame: FrameNode = findTopFrame(this.node);
+    const frameLibrary: Library = getDsLibrary(getData(frame, 'libraries') || []);
```
[the remark this answers is not in the document]
Comment (lynnerang) [reply specter2/src/Identifier.ts:265]: the [] is the fallback for getData, since getData can return a variety of types and I want to be able to specify the fallback for each situation where it's called

File: specter2/src/Identifier.ts [hunk 8]
```diff
@@ -249,37 +249,51 @@ export default class Identifier {
...
-        styleText = `${textToSet}${subtextToSet ? `, ${subtextToSet}` : ''}`;
+      if (text) {
+        // Note: stroke and fill styles both point to the COLOR/PaintStyle details
+        library = styleIds.reduce((acc, id) => {
+          if (hasUnmixedStyle(id)) {
+            const style: BaseStyle = figma.getStyleById(this.node[id]);
+            const staplerLibrary: Library = style && getStaplerData(style)?.library;
+            // in case multiple styles have multiple libraries, let conflicting library
+            // take priority to trigger a UI validation error
+            if ((!acc || acc !== frameLibrary) && staplerLibrary) {
```
[the remark this answers is not in the document]
Comment (lynnerang) [reply specter2/src/Identifier.ts:288]: sure, I'll add that in now

File: specter2/src/Identifier.ts [hunk 9]
```diff
@@ -249,37 +249,51 @@ export default class Identifier {
...
-   * Checks a node for remote styles and concatonates their names.
+   * Checks a node for remote styles and concatonates their names, and checks
+   * for a Stapler assigned library.
    *
    * @kind function
-   * @name getStyleText
+   * @name getStyleDetails
    *
-   * @returns {string} The concatinated style names.
+   * @returns {object} The concatinated style names and the assigned library.
```
[the remark this answers is not in the document]
Comment (lynnerang) [reply specter2/src/Identifier.ts:258]: it's defined in the return value of the function above, but I can re-define it as a variable here as well if that helps

File: specter2/src/App.ts (hunk 4 above)
Comment (lynnerang) [reply specter2/src/App.ts:123]: sure, I'll swap it for your suggestion

File: specter2/src/views/Annotations/Annotations.svelte (hunk 5 above)
Comment (lynnerang) [reply specter2/src/views/Annotations/Annotations.svelte:58]: sure thing, I'll set some vars instead

File: specter2/src/Identifier.ts (hunk 6 above)
Comment (lynnerang) [reply specter2/src/Identifier.ts:282]: I think this has come up before, looked it up just to fact-check myself and it seems to be one of those "matter of opinion" things.  I have a pretty strong preference for this convention in my own code because:

1. everywhere I've worked has had a nested data element and used reduce pretty heavily, and they all opted for this convention (so it's pretty deeply engrained for me)

2. the result of the reduce is already named in the variable that it is assigned to

3. acc refers to the ongoing gathering of that end result, and I find this standard naming convention much easier to see what is representing the looping accumulation across a variety of reduce functions, and naming it something different each time would make that harder imo

File: specter2/src/Identifier.ts (hunk 9 above)
[the remark this answers is not in the document]
Comment (joeyyang) [reply specter2/src/Identifier.ts:258]: yeah, like giving this a type (`Style`, `StyleToCheck`?) and referencing it throughout the file - looking at the ts docs just now it might not be possible to assign a type to a tuple, feel free to ignore this comment if this is the case

[the remark this answers is not in the document]
Comment (lynnerang) [reply specter2/src/Identifier.ts:258]: ohhh I gotcha.  I think this is the only place I combine these into an object like this, was just a way to be able to return two values out of the function
